tests/lib/rules/no-const-assign.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview Tests for no-const-assign rule.
+ * @author Toru Nagashima
+ * @copyright 2015 Toru Nagashima. All rights reserved.
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+var eslint = require("../../../lib/eslint");
+var ESLintTester = require("eslint-tester");
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+var eslintTester = new ESLintTester(eslint);
+eslintTester.addRuleTest("lib/rules/no-const-assign", {
+    valid: [
+        {code: "const x = 0; { let x; x = 1; }", ecmaFeatures: {blockBindings: true}},
+        {code: "const x = 0; function a(x) { x = 1; }", ecmaFeatures: {blockBindings: true}},
+        {code: "const x = 0; foo(x);", ecmaFeatures: {blockBindings: true}},
+        {code: "for (const x in [1,2,3]) { foo(x); }", ecmaFeatures: {blockBindings: true}},
+        {code: "for (const x of [1,2,3]) { foo(x); }", ecmaFeatures: {blockBindings: true, forOf: true}}
+    ],
+    invalid: [
+        {
+            code: "const x = 0; x = 1;",
+            ecmaFeatures: {blockBindings: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "const x = 0; ({x}) = {x: 1};",
+            ecmaFeatures: {blockBindings: true, destructuring: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "const x = 0; ({a: x = 1}) = {};",
+            ecmaFeatures: {blockBindings: true, destructuring: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "const x = 0; x += 1;",
+            ecmaFeatures: {blockBindings: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "const x = 0; ++x;",
+            ecmaFeatures: {blockBindings: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "for (const i = 0; i < 10; ++i) { foo(i); }",
```

Comment:
`for-in`/`for-of` treats its loop variable specially in background if the variable is declared `let`/`const`.
In that case, `for-in`/`for-of` redefine its loop variable on each loop step, not modified.

On V8, this code is valid under strict mode:

```js
for (const a in [1,2,3]) { console.log(a); }
for (const a of [1,2,3]) { console.log(a); }
```

But this code raises a runtime error.

```js
for (const i = 0; i < 3; ++i) { console.log(i); }
```

Reference: 

- [ForIn/OfBodyEvaluation](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset) - 5. b. iv.
- [BindingInstantiation](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-bindinginstantiation)